lightning/src/routing/router.rs

```diff
@@ -351,50 +367,107 @@ pub fn get_route<C: Deref, L: Deref>(our_node_id: &PublicKey, net_graph_msg_hand
+						// hop, if the last hop was provided via a BOLT 11 invoice (though we
+						// should be able to extend it further as BOLT 11 does have feature
+						// flags for the last hop node itself).
+						assert!(res.last().unwrap().pubkey == *target);
+					}
+					if res.last().unwrap().pubkey == *target {
+						break;
 					}
-				} else {
-					// We should be able to fill in features for everything except the last
-					// hop, if the last hop was provided via a BOLT 11 invoice (though we
-					// should be able to extend it further as BOLT 11 does have feature
-					// flags for the last hop node itself).
-					assert!(res.last().unwrap().pubkey == *target);
-				}
-				if res.last().unwrap().pubkey == *target {
-					break;
-				}
 
-				let new_entry = match dist.remove(&res.last().unwrap().pubkey) {
-					Some(hop) => hop.3,
-					None => return Err(LightningError{err: "Failed to find a non-fee-overflowing path to the given destination", action: ErrorAction::IgnoreError}),
-				};
-				res.last_mut().unwrap().fee_msat = new_entry.fee_msat;
-				res.last_mut().unwrap().cltv_expiry_delta = new_entry.cltv_expiry_delta;
-				res.push(new_entry);
+					let new_entry = match dist.remove(&res.last().unwrap().pubkey) {
+						Some(hop) => hop.3,
+						None => {
+							if payment_paths.len() == 0 {
+								// Can't find a first path means there is no point to continue trying.
+								return Err(LightningError{err: "Failed to find a non-fee-overflowing path to the given destination", action: ErrorAction::IgnoreError})
+							} else {
+								break 'path_construction;
+							}
+						}
+					};
+
+
+					if !removable_used_channels.contains(&new_entry.short_channel_id) {
+						let mut removable = true;
+						if last_hops.iter().any(|last_hop| last_hop.short_channel_id==new_entry.short_channel_id) {
+							removable = false;
+						}
+
+						if let Some(hops) = first_hops {
+							if hops.iter().any(|first_hop| first_hop.short_channel_id==Some(new_entry.short_channel_id)) {
+								removable = false;
+							}
+
+						}
+
+						if removable {
+							removable_used_channels.push(new_entry.short_channel_id);
+						}
+					}
+
+					res.last_mut().unwrap().fee_msat = new_entry.fee_msat;
+					res.last_mut().unwrap().cltv_expiry_delta = new_entry.cltv_expiry_delta;
+					res.push(new_entry);
+				}
+				res.last_mut().unwrap().fee_msat = final_value_msat;
+				res.last_mut().unwrap().cltv_expiry_delta = final_cltv;
+				payment_paths.insert(res);
+				break 'path_construction;
+			}
+			match network.get_nodes().get(&pubkey) {
+				None => {},
+				Some(node) => {
+					let mut channels_to_avoid = HashSet::new();
+					if removable_used_channels.len() > 0 {
+						// This sampling can be improved when we update the rand crate.
+						let mut rng = thread_rng();
+						let used_channels_indexes = Range::new(0, removable_used_channels.len());
+						// We avoid just one random channel currently, this can be improved,
```

Comment:
I'm kinda confused what we're doing here. Can you elaborate a bit more on the algorithm used here?

Reply:
It's actually very simple.
1. First A* pass, we compute a regular path (fully quit if not found). Store all used channels but first/last hops in `removable_used_channels`.
2. From `removable_used_channels` take 1 random channel and put it in `channels_to_avoid`
3. Do second A* pass, but skip those channels belonging to `channels_to_avoid` (currently just 1 channel always). If found, add used channels to `removable_used_channels`
4. Repeat 2-3 until found enough paths or wasted all tries.

(Non-first A* passes *can* currently hit a duplicate paths with each other, but that's probably fine)

As I mentioned in the original post, there are a bunch of improvements I'm planning to make (for instance, 2 can "avoid" multiple paths).